File: python/priorityQueue/nAryHeap.py
class NAryHeap:
    def __init__(self, n):
        self.n = n
        self.H = [None] * 1
        self.size = -1
        self.maxSize = 1

    def toString(self):
        return ','.join([str(self.H[i]) for i in range(self.size + 1)])

    def _parent(self, i):
        return (i - 1) // self.n
    
    def _leftChild(self, i):
        return i * self.n + 1

    def _rightChild(self, i):
        return i * self.n + self.n

    def _swap(self, i, j):
        self.H[i], self.H[j] = self.H[j], self.H[i]

    def _shiftUp(self, i):
        while i > 0 and self.H[i] > self.H[self._parent(i)]:
            p = self._parent(i)
            self._swap(i, p)
            i = p

    def _shiftDown(self, i):
        maxIdx = i

        for cIdx in range(self._leftChild(i), self._rightChild(i) + 1):
            if cIdx <= self.size and self.H[maxIdx] < self.H[cIdx]:
                maxIdx = cIdx

        if i != maxIdx:
            self._swap(i, maxIdx)
            self._shiftDown(maxIdx)

    def insert(self, p):
        if self.size == self.maxSize - 1:
            newH = [None] * (self.maxSize * self.n)
            for i in range(self.maxSize):
                newH[i] = self.H[i]
            self.H = newH
            self.maxSize = self.maxSize * self.n

        self.size += 1
        self.H[self.size] = p
        self._shiftUp(self.size)

    def extractMax(self):
        if self.size == -1:
            raise IndexError
        result = self.H[0]
        self.H[0] = self.H[self.size]
        self.size -= 1
        self._shiftDown(0)

        return result

    def remove(self, i):
        if i < 0 or i > self.size:
            raise IndexError
        self.H[i] = self.H[0] + 1
        self._shiftUp(i)
        self.extractMax()

    def changePriority(self, p, i):
        if i < 0 or i > self.size:
            raise IndexError
        oldp = self.H[i]
        self.H[i] = p
        if oldp < p:
            self._shiftUp(i)
        else:
            self._shiftDown(i)

File: java/src/priorityQueue/nAryHeap.java
package priorityQueue;

public class nAryHeap {
    private int n;

    private int[] H = new int[1];

    private int size = -1;

    private int maxSize = 1;


    public nAryHeap() {}

    public nAryHeap(int n) {
        this.n = n;
    }

    @Override
    public String toString() {
        StringBuffer sb = new StringBuffer();
        for (int i = 0 ; i <= this.size; i++) {
            sb.append(this.H[i]);
            sb.append(",");
        }

        return sb.toString();
    }


    private int parent(int i) {
        return (int) Math.floor((i - 1) / this.n);
    }

    private int leftChild(int i) {
        return i * this.n + 1;
    }

    private int rightChild(int i) {
        return i * this.n + this.n;
    }

    private void swap(int i, int j) {
        int tmp = this.H[i];
        this.H[i] = this.H[j];
        this.H[j] = tmp;
    }

    private void shiftUp(int i) {
        int p = 0;

        while(i > 0 && this.H[i] > this.H[this.parent(i)]) {
            p = this.parent(i);
            this.swap(i, p);
            i = p;
        }
    }

    private void shiftDown(int i) {
        int maxIdx = i;

        for (int cIdx = this.leftChild(i); cIdx <= this.rightChild(i); cIdx++)
            if (cIdx <= this.size && this.H[maxIdx] < this.H[cIdx])
                maxIdx = cIdx;

        if (i != maxIdx) {
            this.swap(i, maxIdx);
            this.shiftDown(maxIdx);
        }
    }

    public void insert(int p) {
        if (this.size == this.maxSize - 1) {
            int[] newH = new int[this.maxSize * this.n];

            for (int i = 0; i <= this.size; i++)
                newH[i] = this.H[i];

            this.H = newH;
            this.maxSize = this.maxSize * this.n;
        }

        this.H[++this.size] = p;
        this.shiftUp(this.size);
    }

    public int extractMax() {
        if (this.size == -1)
            throw new IndexOutOfBoundsException();

        int result = this.H[0];
        this.H[0] = this.H[this.size];
        this.size--;
        this.shiftDown(0);

        return result;
    }

    public void remove(int i) {
        if (i < 0 || i > this.size)
            throw new IndexOutOfBoundsException();

        this.H[i] = this.H[0] + 1;
        this.shiftUp(i);
        this.extractMax();
    }

    public void changePriority(int p, int i) {
        if (i < 0 || i > this.size)
            throw new IndexOutOfBoundsException();

        int oldp = this.H[i];
        this.H[i] = p;
        if (oldp < p)
            this.shiftUp(i);
        else
            this.shiftDown(i);
    }


}
